src/Microsoft.AspNetCore.Blazor.Browser/Hosting/BrowserHostBuilder.cs

```diff
@@ -0,0 +1,103 @@
+    internal class BrowserHostBuilder : IBrowserHostBuilder
+    {
+        private List<Action<BrowserHostBuilderContext, IServiceCollection>> _configureServicesActions = new List<Action<BrowserHostBuilderContext, IServiceCollection>>();
+        private bool _hostBuilt;
+        private BrowserHostBuilderContext _BrowserHostBuilderContext;
+        private IBrowserHostingEnvironment _hostingEnvironment;
+        private IServiceProvider _appServices;
+
+        /// <summary>
+        /// A central location for sharing state between components during the host building process.
+        /// </summary>
+        public IDictionary<object, object> Properties { get; } = new Dictionary<object, object>();
+
+        /// <summary>
+        /// Adds services to the container. This can be called multiple times and the results will be additive.
+        /// </summary>
+        /// <param name="configureDelegate"></param>
+        /// <returns>The same instance of the <see cref="IBrowserHostBuilder"/> for chaining.</returns>
+        public IBrowserHostBuilder ConfigureServices(Action<BrowserHostBuilderContext, IServiceCollection> configureDelegate)
+        {
+            _configureServicesActions.Add(configureDelegate ?? throw new ArgumentNullException(nameof(configureDelegate)));
+            return this;
+        }
+
+        /// <summary>
+        /// Run the given actions to initialize the host. This can only be called once.
+        /// </summary>
+        /// <returns>An initialized <see cref="IBrowserHost"/></returns>
+        public IBrowserHost Build()
+        {
+            if (_hostBuilt)
+            {
+                throw new InvalidOperationException("Build can only be called once.");
+            }
+            _hostBuilt = true;
+
+            CreateHostingEnvironment();
+            CreateBrowserHostBuilderContext();
+            CreateServiceProvider();
+
+            return _appServices.GetRequiredService<IBrowserHost>();
+        }
+
+        private void CreateHostingEnvironment()
+        {
+            _hostingEnvironment = new BrowserHostingEnvironment();
+        }
+
+        private void CreateBrowserHostBuilderContext()
+        {
+            _BrowserHostBuilderContext = new BrowserHostBuilderContext(Properties)
+            {
+                HostingEnvironment = _hostingEnvironment,
+            };
+        }
+
+        private void CreateServiceProvider()
+        {
+            var services = new ServiceCollection();
+            services.AddSingleton(_hostingEnvironment);
+            services.AddSingleton(_BrowserHostBuilderContext);
+            services.AddSingleton<IBrowserHost, BrowserHost>();
+            services.AddSingleton<IJSRuntime, MonoWebAssemblyJSRuntime>();
+
+            services.AddSingleton<IUriHelper, BrowserUriHelper>();
+            services.AddSingleton<HttpClient>(s =>
+            {
+                // Creating the URI helper needs to wait until the JS Runtime is initialized, so defer it.
```

Comment:
Calling `GetBaseUri` on init was convenient when I did it the first time, but I appreciate it complicates things here. It will also complicate things in the server case.

No need to change this now, but later, we should consider having the base URI and current URI be something that gets passed into the .NET code as part of the "please boot up now" invocation.

Reply:
Yeah that would simplify things a good deal.